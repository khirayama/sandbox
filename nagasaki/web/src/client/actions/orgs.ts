import { Action } from 'redux';
import { State } from '../reducers/orgs';

export type Actions = FetchRepos | FetchReposSuccess | FetchReposFailure | ResetOrgs;

export interface FetchRepos extends Action<'FETCH_REPOS'> {
  payload: {
    org: string;
  };
}

export const fetchRepos = (org: FetchRepos['payload']['org']): FetchRepos => ({
  type: 'FETCH_REPOS',
  payload: {
    org
  }
});

export interface FetchReposSuccess extends Action<'FETCH_REPOS_SUCCESS'> {
  payload: {
    name: string;
    repos: State['repos'];
  };
}

export const fetchReposSuccess = ({
  name,
  repos
}: FetchReposSuccess['payload']): FetchReposSuccess => ({
  type: 'FETCH_REPOS_SUCCESS',
  payload: {
    name,
    repos
  }
});

export interface FetchReposFailure extends Action<'FETCH_REPOS_FAILURE'> {
  payload: {
    error: Error;
  };
}

export const fetchReposFailure = (error: Error): FetchReposFailure => ({
  type: 'FETCH_REPOS_FAILURE',
  payload: {
    error
  }
});

export interface ResetOrgs extends Action<'RESET_ORGS'> {}

export const resetOrgs = () => ({
  type: 'RESET_ORGS'
});
