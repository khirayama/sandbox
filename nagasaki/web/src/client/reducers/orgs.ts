import { Actions } from '../actions/orgs';

export interface State {
  name: string;
  repos: Array<{
    url: string;
    name: string;
    language: string;
    forksCount: number;
    issuesCount: number;
    stargazersCount: number;
    watchersCount: number;
  }>;
  isFetching: boolean;
}

export const initialState: State = {
  name: '',
  repos: [],
  isFetching: false
};

export const reducer = (state: State = initialState, action: Actions): State => {
  switch (action.type) {
    case 'FETCH_REPOS':
      return { ...state, isFetching: true };
    case 'FETCH_REPOS_SUCCESS':
      return { ...state, isFetching: false, ...action.payload };
    case 'RESET_ORGS':
      return initialState;
    default:
      return state;
  }
};
