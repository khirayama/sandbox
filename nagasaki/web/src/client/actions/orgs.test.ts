import * as actions from './orgs';

test('should create an action to fetch repositories', () => {
  expect(actions.fetchRepos('foo')).toMatchSnapshot();
});

test('should create a successful action to fetch repositories', () => {
  expect(
    actions.fetchReposSuccess({
      name: 'foo',
      repos: []
    })
  ).toMatchSnapshot();
});

test('should create a failure action to fetch repositories', () => {
  expect(actions.fetchReposFailure(new Error())).toMatchSnapshot();
});

test('should create an action to reset state', () => {
  expect(actions.resetOrgs()).toMatchSnapshot();
});
