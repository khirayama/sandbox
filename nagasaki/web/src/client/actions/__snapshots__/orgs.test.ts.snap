// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should create a failure action to fetch repositories 1`] = `
Object {
  "payload": Object {
    "error": [Error],
  },
  "type": "FETCH_REPOS_FAILURE",
}
`;

exports[`should create a successful action to fetch repositories 1`] = `
Object {
  "payload": Object {
    "name": "foo",
    "repos": Array [],
  },
  "type": "FETCH_REPOS_SUCCESS",
}
`;

exports[`should create an action to fetch repositories 1`] = `
Object {
  "payload": Object {
    "org": "foo",
  },
  "type": "FETCH_REPOS",
}
`;

exports[`should create an action to reset state 1`] = `
Object {
  "type": "RESET_ORGS",
}
`;
